_posts/2021-09-06-connector-table-sql-api-part2.md

```diff
@@ -0,0 +1,519 @@
+
+
+# Add configuration options - server information and credentials
+
+In order to connect to your IMAP server, you will need at least the following:
+
+- hostname (of the mail server)
+- port number
+- username
+- password
+
+You will start by creating a class to encapsulate the configuration options. You will make use of [Lombok](https://projectlombok.org) to help with some boilerplate code. By adding the `@Data` and `@SuperBuilder` annotations, Lombok will generate these for all the fields of the immutable class. 
+
+```java
+import lombok.Data;
+import lombok.experimental.SuperBuilder;
+import javax.annotation.Nullable;
+import java.io.Serializable;
+
+@Data
+@SuperBuilder(toBuilder = true)
+public class ImapSourceOptions implements Serializable {
+    private static final long serialVersionUID = 1L;
+
+    private final String host;
+    private final @Nullable Integer port;
+    private final @Nullable String user;
+    private final @Nullable String password;
+}
+```
+
+Now you can add an instance of this class to the `ImapSource` and `ImapTableSource` classes previously created (in part one) so it can be used there. Take note of the column names with which the table has been created. This will help later.
+
+
+<div class="note">
+  <h5>Hint</h5>
+  <p>The column names would be "subject" and "content" with the SQL executed in part one:</p>
+  <pre><code class="language-sql">CREATE TABLE T (subject STRING, content STRING) WITH ('connector' = 'imap');</code></pre>
+</div>
+
+
+```java
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.flink.table.data.RowData;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ImapSource extends RichSourceFunction<RowData> {
+    private final ImapSourceOptions options;
+    private final List<String> columnNames;
+
+    public ImapSource(
+        ImapSourceOptions options, 
+        List<String> columnNames
+    ) {
+        this.options = options;
+        this.columnNames = columnNames.stream()
+            .map(String::toUpperCase)
+            .collect(Collectors.toList());
+    }
+
+    // ...
+}
+```
+
+```java
+import org.apache.flink.table.connector.source.DynamicTableSource;
+import org.apache.flink.table.connector.source.ScanTableSource;
+import java.util.List;
+
+public class ImapTableSource implements ScanTableSource {
+
+    private final ImapSourceOptions options;
+    private final List<String> columnNames;
+
+    public ImapTableSource(
+        ImapSourceOptions options,
+        List<String> columnNames
+    ) {
+        this.options = options;
+        this.columnNames = columnNames;
+    }
+
+    // …
+
+    @Override
+    public ScanRuntimeProvider getScanRuntimeProvider(ScanContext ctx) {
+        final ImapSource source = new ImapSource(options, columnNames);
+        return SourceFunctionProvider.of(source, true);
```

Comment:
are we mentioning that our source is bounded and the steps required to make it unbounded?

Reply:
I guess this is actually wrong here; the source should be unbounded at this point. I'll fix it and point it out.